### PR TITLE
Support of optional nested-nested types reading

### PR DESCRIPTION
**Pull request by oldukhno**
Fixed a bug and added unit tests to cover nested-nested types writing/reading when nested is list and nested-nested is either optional parquet list or parquet map with a required leaf field.

### CODE REVIEW

**Review by marcin-krystianc — COMMENTED** [review]
I think that the changes are good. There is a formatting issue and a typo in the name of the test method though.

File: csharp.test/TestMaps.cs [hunk 1]
```diff
@@ -48,6 +49,110 @@ public static void CanRoundtripRequiredMaps()
             DoRoundtripTest(false, keys, values);
         }
 
+        /// <summary>
+        /// This checks that values written to nested-nested required key field of 
+        /// a nested-nested optional map can be read back.
+        /// </summary>
+        [Test]
+        public static void TestNestedNestedOptinalMapWithRequiredKey()
```
[the remark this answers is not in the document]
Comment (oldukhno) [reply csharp.test/TestMaps.cs:57]: thanks for catching this

**Comment by oldukhno** [comment]
@marcin-krystianc , could you please share the plan to release new nuget having the Nested<> type?

**Comment by adamreeve** [comment]
Thanks for the fix and new tests @oldukhno. If I'm understanding this correctly, this bug is specific to string values as in  `LogicalRead.GetConverter` we always return a conversion function that uses the definition levels for string values but for other types we use different conversion functions depending on the nullability of the leaf-level type. So an alternative approach would be to provide a different string conversion function there for required string values. But your approach is nice as it's not type specific.

The nested type support seems like a big enough change that it should be in a major release even though it should be mostly backwards compatible. We don't really have a release timeline planned out but aim to release new major versions corresponding to Arrow major releases, and we're lagging behind a bit now as Arrow has released 10.0.0 and we're only at 8.0.0 so we are probably overdue a new major release. We usually release beta versions to NuGet before making a final release so could possibly get a 9.0.0 or 10.0.0 beta out soon. I'll look into this and get back to you.

Also FYI the Windows build failure is due to https://github.com/actions/runner-images/issues/6629 which should be resolved soon.

**Review by adamreeve — APPROVED** [review]

**Comment by oldukhno** [comment]
> Thanks for the fix and new tests @oldukhno. If I'm understanding this correctly, this bug is specific to string values as in `LogicalRead.GetConverter` we always return a conversion function that uses the definition levels for string values but for other types we use different conversion functions depending on the nullability of the leaf-level type. So an alternative approach would be to provide a different string conversion function there for required string values. But your approach is nice as it's not type specific.
> 
> The nested type support seems like a big enough change that it should be in a major release even though it should be mostly backwards compatible. We don't really have a release timeline planned out but aim to release new major versions corresponding to Arrow major releases, and we're lagging behind a bit now as Arrow has released 10.0.0 and we're only at 8.0.0 so we are probably overdue a new major release. We usually release beta versions to NuGet before making a final release so could possibly get a 9.0.0 or 10.0.0 beta out soon. I'll look into this and get back to you.
> 
> Also FYI the Windows build failure is due to [actions/runner-images#6629](https://github.com/actions/runner-images/issues/6629) which should be resolved soon.

Even better approach would be not posting the definition levels along with the non-nullable column values. Yet as I got it, there is a constraint on the apache arrow side which requires the number of definition levels match the number of values in the column. In any case, the code on the reader side does the necessary trick.

Looking forward to getting the change into master and getting an estimate on the new nuget package release.

Thanks

**Comment by adamreeve** [comment]
Hi @oldukhno, I've made a PR to upgrade to Arrow 10.0.1 (#321) and the current plan is to release a 10.0.0-beta soon with that change.

**Comment by adamreeve** [comment]
Hi @oldukhno, this change has been released in version 10.0.1-beta1

**Pull request merged**